### PR TITLE
ASCII art expansion for croatoan

### PR DESCRIPTION
**Pull request by veddox**
**Description:** This pull request forms an ncurses extension to generate simple ASCII art shapes. It features the class `shape`, which consists in the main of a list of coordinates, and a function to render a shape onto a croatoan window. It also includes functions to generate basic shapes, such as lines, rectangles or circles, and combine these into more complex shapes. For an example of what the class can do, the test `t29a` is provided.

**Rationale:** I've been using croatoan to create a game, and though I don't need shapes at the moment, I thought it would be a neat feature to have available for future use. I'm not sure whether you consider it suitable or worthwhile to have such an extension in the main package, but I at least wanted to make the offer. If you decline, I might expand my work so far into its own package and try and get it published separately ;-)

### CODE REVIEW

**Comment by McParen** [comment]
Hello and thanks for your interest in using croatoan and for your contribution. I certainly do think that adding new neat features into the main library is a worthwhile effort. As time permits, I will review your code and merge it in. I hope you agree with publishing your code under the MIT license. I also hope you agree with slight modifications from my part to make the code fit with the rest of the croatoan code. Thanks!

**Comment by veddox** [comment]
Hello! MIT is my preferred license anyway, so no problem with that. I also don't mind any changes to the code - I tried to keep the style similar to the rest of the package, but obviously there will still be differences.

And thank you for creating croatoan in the first place; I've much appreciated it and found it a pleasure to work with!

File: source/shape.lisp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+			(dolist (c (.coordinates s))
+				(unless (member c (.coordinates shp) :test #'equal)
+					(setf (.coordinates shp)
+						(append (.coordinates shp)
+							(list (list (first c) (second c))))))))))
+
+
+;;; Create various basic shapes
+
+(defun line (y0 x0 y1 x1 &key char)
```
Comment (veddox) [inline source/shape.lisp]: This function looks like it ought to pretty simple, but was actually really tricky to get exactly right... That's why it's ended up as big as it is now. I tried to comment it sufficiently to keep it understandable. But if you can think of a more elegant way of drawing a line that still takes care of all the edge cases, you are welcome to replace this function :-)

File: source/shape.lisp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+		(dolist (s shapes shp)
+			(dolist (c (.coordinates s))
+				(unless (member c (.coordinates shp) :test #'equal)
+					(setf (.coordinates shp)
+						(append (.coordinates shp)
+							(list (list (first c) (second c))))))))))
+
+
+;;; Create various basic shapes
+
```
Comment (veddox) [inline source/shape.lisp]: This function looks like it ought to pretty simple, but was actually really tricky to get exactly right... That's why it's ended up as big as it is now. I tried to comment it sufficiently to keep it understandable. But if you can think of a more elegant way of drawing a line that still takes care of all the edge cases, you are welcome to replace this function :-)

File: source/shape.lisp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+			 (coords (.coordinates shape) (cdr coords))
+			 (y (first (car coords)) (first (car coords)))
+			 (x (second (car coords)) (second (car coords))))
+		((null coords))
+		(setf y (+ y0 y) x (+ x0 x))
+		(when squarify (setf x (* 2 x)))
+		(unless (or (minusp y) (minusp x) (>= y (.height win)) (>= y (.width win)))
+			(add-char win (.simple-char c) :y y :x x
+				:attributes (.attributes c) :color-pair (.color-pair c)))))
+
```
Comment (veddox) [inline source/shape.lisp]: Just realized that `delete-shape` doesn't have a `squarify` option like `draw-shape` has. That would need to be added. (Or perhaps we don't need `delete-shape` at all and can just get rid of it?)

**Pull request closed without merging**